### PR TITLE
Add logo preview to creator settings form

### LINKED ISSUE #15348: Creator Settings: Add a logo preview to the creator settings form
## Description

Add a logo preview when a creator uploads a logo in the creator settings form.

image

Use the [FileReader](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) API to load the image a creator selects (not uploads) to render the logo preview

Estimate:	1 day

### LINKED ISSUE #15350: Creator Settings: Add validation for logo
## Description

The logo a creator uploads has to have a minimum size of 300px x 300px.

Use the [FileReader](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) API to validate the image  dimensions.

image

@anujbhavsar96 and @Ridhwana, I asked this in the Figma, but where would you like the validation message? Inline close to the input/logo preview or a dynamic flash error message?

Esitmate: 1.25 days

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR adds a logo preview so that a Forem creator can see what it looks like before uploading the file. Note that the original design had the preview logo to the left of the Change Logo button. I opted to put it on the right to avoid a [cummulative layout shift](https://web.dev/cls/) (CLS).

CleanShot 2021-11-18 at 10 48 32

Also note, that we are keeping the default styling for the input of type file at the moment.

## Related Tickets & Documents

Closes #15348
Closes #15350

## QA Instructions, Screenshots, Recordings

1. Ensure you're logged out from the Forem instance.
2. Set up the database to be in a state where you are about to create the Forem creator role

```ruby
# Enable the feature flag
FeatureFlag.enable(:creator_onboarding)

# Delete all users 
User.destroy_all

# Set `waiting_on_first_user` to be true
Settings::General.waiting_on_first_user = true

# If you've logged in before set the mascot id to nil
Settings::General.mascot_user_id = nil
```

3. Click on the login link. Since the feature flag is enabled, you'll be redirected to create a new Forem creator, i.e. first account.
4. Fill in the necessary information: name, email, password and submit the form.
5. On the next page select an image for the logo.
6. Notice that it loads a preview of the image.

https://user-images.githubusercontent.com/833231/142430313-48f2ba2d-0fdf-4d20-a63e-78b9d444a700.mp4

7. Note that a Forem creator can keep changing the logo until they're happy and after each change, the logo preview will update.

8. Fill out the rest of the information in the form, community name, brand colour etc. and click finish.
9. The form submits successfully.

### UI accessibility concerns?

N/A

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [x] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

Loading preview spinner

### CODE REVIEW

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [inline app/views/admin/creator_settings/_form.html.erb:11]: This is no longer the label. There is now a label that looks like a button hiding the actual input of type file. It didn't make sense to make the old label a subheading, so I'm not sure what to use here from a semantic standpoint. Maybe you have a suggestion @aitchiss.

File: jsconfig.json
Comment (nickytonline) [inline jsconfig.json:8]: This is so VS Code recognizes the alias used in webpack.

File: config/webpack/environment.js
Comment (nickytonline) [inline config/webpack/environment.js:29]: The [webpack alias](https://webpack.js.org/configuration/resolve/#resolvealias) for controllers.

File: app/javascript/.eslintrc.js
Comment (nickytonline) [inline app/javascript/.eslintrc.js:31]: This let's eslint know where to find `@controllers/*` files.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [inline app/views/admin/creator_settings/_form.html.erb:23]: I moved this around, because it looks like if there is a logo, it will load it. I took the `<div />` out of the `if` so I can use the same div for dynamic previews.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [inline app/views/admin/creator_settings/_form.html.erb:18]: Now only png and SVG files are permitted.

image

File: app/javascript/packs/base.jsx
Comment (nickytonline) [inline app/javascript/packs/base.jsx:148]: This pack file is loaded even in the creator settings page. I thought of creating a separate one, but it means adding another conditional in how we load pack files. With this approach, it's a quick check on the route and then the controller is dynamically loaded.

File: app/javascript/packs/base.jsx
Comment (nickytonline) [inline app/javascript/packs/base.jsx:166]: TODO: I need to look @citizen428's PR that added the routes gem for frontend code instead of using a hard-coded value here.

**Comment by nickytonline** [comment]
@ludwiczakpawel, you'll notice the upload logo label for the input does get focus, but doesn't get a focus state because it's a label. I didn't see anything in our CSS or maybe I missed it, but basically we need styles for `input[type="file"] + label`. This will probably be a common pattern for styling inputs for uploading.

**Comment by ludwiczakpawel** [comment]
@nickytonline i haven't looked at the code but what if you used `focus-within` either on wrapper OR you could put input inside label and still try to use :focus-within?

File: app/views/admin/creator_settings/new.html.erb
Comment (nickytonline) [inline app/views/admin/creator_settings/new.html.erb:6]: Boostrapping the CreatorSettings Stimulus controller.

File: config/webpack/environment.js
Comment (Ridhwana) [reply config/webpack/environment.js:29]: @nickytonline is this config so that we can now use Stimulus controllers on the user facer side of the app (like the Creator Settings Controller)?

Comment (nickytonline) [reply config/webpack/environment.js:29]: No, this config just allows us to use an alias is all. It's more to avoid having to do e.g. `../../controllers/some-controller.js` vs. `@controllers/some-controller.js`.

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [reply app/javascript/packs/base.jsx:166]: Probably just need to use: `window.Forem.routes.new_admin_creator_setting_path()`

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [reply app/javascript/packs/base.jsx:148]: ok, actually this seems to be how we load the stimulus controller to the user facing side.

Comment (Ridhwana) [reply app/javascript/packs/base.jsx:148]: Do you have any long term plans for this, if we wanted to load another stimulus controller on the user facing side, would we generalize this function or would there be a new async loadXYZ for each controller?

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [reply app/javascript/packs/base.jsx:166]: I think it's something like this

```
import { newAdminCreatorSettingPath } from 'routes.js.erb';
console.log(newAdminCreatorSettingPath());
```

Comment (Ridhwana) [reply app/javascript/packs/base.jsx:166]: (blocking): Also, when a user directs from the login screen the url will be `http://localhost:3000/admin/creator_settings/new?referrer=http://localhost:3000/` hence the condition `document.location.href.endsWith('/admin/creator_settings/new')` will evaluate to `false`.

File: app/javascript/packs/base.jsx
Comment (nickytonline) [reply app/javascript/packs/base.jsx:148]: > Do we have any long term plans for this - it seems to be the first time we use a stimulus controller in a non-admin layout view.  
> 
> 
> 
> How does it scale? If we wanted to load another stimulus controller here, would we generalize this function or would there be a new async loadXYZ for each controller? 
> 
> 
> 
> 
> 
> Also, if we wanted to use the same stimulus controller on the admin, would we be able to do that? I can see this being the case for image upload, we'd use the same controller from creator settings on the admin for for image upload.

It's being loaded dynamically so it does not affect page loads of the client-facing portion of the application aside from a quick Bboolean check on the url.

As you mention, it is the first time we do this. We could probably take a page out of the rails handbook and use the route to determine which controller to load on the client-side. That way we could have one function to dynamically load a Stimulus controller if one exists. Something along those lines. We can discuss it further as I'm 100% open to suggestions/I haven't thought about it too much yet. 😎

File: app/views/admin/creator_settings/_form.html.erb
Comment (aitchiss) [reply app/views/admin/creator_settings/_form.html.erb:11]: This is how I approached the refactor to the ImageUploader button too, as like you say you can't style the file input button (maybe one day?!)

I would tend to agree that "Logo" probably doesn't need to be associated with the file input, since the label text already says "Upload logo" and is pretty self explanatory. We should link the description with the input though by using `aria-describedby="logo-subtitle"` on line 20, just to get that extra info passed along to screen reader users

File: app/views/admin/creator_settings/_form.html.erb
Comment (aitchiss) [inline app/views/admin/creator_settings/_form.html.erb:13]: For this to work, it would need to be changed to 
```suggestion
    <span class="crayons-field__required crayons-hover-tooltip" data-tooltip="This will set the logo for your Forem"></span>
```

and I think the `aria-describedby` wouldn't help in this situation, as a hover only tooltip isn't accessible to keyboard navigation (TBH I wonder if we need it all - it feels to me like all the necessary info is already presented to the user)

File: app/views/admin/creator_settings/_form.html.erb
Comment (aitchiss) [reply app/views/admin/creator_settings/_form.html.erb:23]: I can't decide if this should be `aria-hidden` or if the images should just have alt text confirming e.g. the filename or "currently selected logo" or smth.

I'm just thinking about it from the POV of not all screen reader users being visually impaired and wondering what this block is

File: app/javascript/controllers/creator_settings_controller.js
Comment (aitchiss) [inline app/javascript/controllers/creator_settings_controller.js:42]: I think this is already set on line 21?

File: app/assets/stylesheets/components/buttons.scss
Comment (aitchiss) [inline app/assets/stylesheets/components/buttons.scss:136]: This also fixes https://github.com/forem/forem/issues/15234 🎉

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb:23]: Yeah, to me it seem like a visual enhancement only since the file selection should be enough, but im flip flopping still.

Comment (aitchiss) [reply app/views/admin/creator_settings/_form.html.erb:23]: I guess an empty `alt` tag on the image would do the same job too and avoid using ARIA?

I think yeah there's maybe not much to describe since it's just the current file input value really.

File: app/views/admin/creator_settings/_form.html.erb
Comment (ludwiczakpawel) [reply app/views/admin/creator_settings/_form.html.erb]: Any reason for NOT supporting GIF or JPG?

File: app/assets/stylesheets/components/buttons.scss
Comment (nickytonline) [reply app/assets/stylesheets/components/buttons.scss:136]: I'm going to put this in a separate PR to unblock that issue, and it's really a global fix.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb]: I'm going to go ahead and remove it.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb]: We can. I thought the requirement was SVG or PNG.

**Comment by nickytonline** [comment]
> @nickytonline i haven't looked at the code but what if you used `focus-within` either on wrapper OR you could put input inside label and still try to use :focus-within?

Yeah, that's what I ended up going with. See #15368.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb]: I flip-flopped on this again. Now there is alt text that says `preview of logo selected` on the image and I updated the E2E test to check for it. I think it makes sense to let them know the image preview is there, but let me know what you think.

File: app/javascript/packs/base.jsx
Comment (nickytonline) [inline app/javascript/packs/base.jsx:166]: @Ridhwana, this works better so  that there's no need to worry about whether or not a querystring is present.

File: app/views/admin/creator_settings/_form.html.erb
Comment (aitchiss) [reply app/views/admin/creator_settings/_form.html.erb]: I think that's grand. Could potentially be helpful for screen reader users who can perceive the image well

File: app/views/admin/creator_settings/_form.html.erb
Comment (ludwiczakpawel) [reply app/views/admin/creator_settings/_form.html.erb]: Yea, that probably wasn't very clear. I think we should enable all sensible format: SVG, PNG, JP(E)G and GIF. Obviously SVG is recommended, and PNG is, I guess 2nd recommended... :D

Comment (aitchiss) [reply app/views/admin/creator_settings/_form.html.erb]: I'd maybe suggest we don't allow GIF? It would be pretty bad for accessibility to have an animated logo when we don't currently have any settings to disable that. We already have (or had, at least) plans to disable this for user avatars - https://github.com/forem/rfcs/issues/125#issuecomment-786926070

File: app/javascript/controllers/creator_settings_controller.js
Comment (aitchiss) [inline app/javascript/controllers/creator_settings_controller.js:55]: I'm seeing an error in the console in the case where a user opens the file picker, but instead of choosing a file, clicks "Cancel":
error executing 'readDataAsURL'

I think that made me get into this weird state where I had:

- picked an image, and saw the preview as expected
- opened the image picker again, and clicked cancel
- couldn't submit the form, but no error message was visible

https://user-images.githubusercontent.com/20773163/141983453-af40b468-5624-4791-ad45-0d8eb8200427.mp4

File: app/views/admin/creator_settings/_form.html.erb
Comment (aitchiss) [inline app/views/admin/creator_settings/_form.html.erb:13]: I haven't been able to trigger this so far - I think the span is really small so I can't manage to hover it with the mouse

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [reply app/javascript/packs/base.jsx:166]: Looks good @nickytonline , also shall we user routes.js?

```
import { newAdminCreatorSettingPath } from 'routes.js.erb';

if (document.location.pathname === newAdminCreatorSettingPath()) {
  loadCreatorSettings();
}
```

File: cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js
Comment (aitchiss) [inline cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js:70]: One thing I noticed is that although this field is required, there's no visual indication of that other than the button focusing after submit click:

https://user-images.githubusercontent.com/20773163/141985225-128ac1eb-ba80-406f-9902-950dc4c63c6e.mp4

If every field is required I wonder if we should consider adding a note on the page that says that (it's just a thought 🤷‍♀️ )

**Review by aitchiss — COMMENTED** [review]
Added a few questions in the comments, but one other thing I was wondering is whether under this PR I should be expecting my selected logo to show up in the header after the form submits, or if that's a later PR?

At the minute I always just see the image file name (as text):
header showing a cleanshot file name where the logo would normally be

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [reply app/javascript/packs/base.jsx:166]: Looks good @nickytonline , also shall we user routes.js?

```
import { newAdminCreatorSettingPath } from 'routes.js.erb';

if (document.location.pathname === newAdminCreatorSettingPath()) {
  loadCreatorSettings();
}
```

File: app/javascript/packs/base.jsx
Comment (Ridhwana) [inline app/javascript/packs/base.jsx:162]: @nickytonline will you be adding this to this PR?

File: app/javascript/controllers/creator_settings_controller.js
Comment (nickytonline) [reply app/javascript/controllers/creator_settings_controller.js:55]: Good ctach. I hadn't considered the cancel. I'll get that sorted.

File: cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js
Comment (nickytonline) [reply cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js:70]: Ahh, another good catch. This is because we're using default HTML validation and since the actual input is not visible, its validation message never appears. Here it is when it's visible. 

image

I think the note makes sense, or I can see how I can make it appear. 🤔

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb:13]: I fixed this. Here's a recording.

https://user-images.githubusercontent.com/833231/142026594-162d991a-58ed-41d5-926f-be2c41a434b1.mov

**Comment by nickytonline** [comment]
> Added a few questions in the comments, but one other thing I was wondering is whether under this PR I should be expecting my selected logo to show up in the header after the form submits, or if that's a later PR?
>
> At the minute I always just see the image file name (as text):
header showing a cleanshot file name where the logo would normally be

@aitchiss, this is something we need to sort out, but it's out of the scope of previewing a logo that's about to be uploaded.

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb]: I spoke with @Ridhwana. @aitchiss and @ludwiczakpawel, I added JPG, but we're opting out of GIF for now to avoid the potential for someone uploading an animated GIF.

File: app/javascript/controllers/creator_settings_controller.js
Comment (nickytonline) [reply app/javascript/controllers/creator_settings_controller.js]: Removed it.

File: app/javascript/packs/base.jsx
Comment (nickytonline) [reply app/javascript/packs/base.jsx]: As discussed, I'm going to remove this. It's being a little too over defensive in terms of coding.

File: app/assets/stylesheets/setup-mode.scss
Comment (nickytonline) [inline app/assets/stylesheets/setup-mode.scss:34]: This is using the [:focus](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus) pseudo-class along with the [:invalid](https://developer.mozilla.org/en-US/docs/Web/CSS/:invalid) pseudo-class, so that we can handle the validation message visibility without JavaScript.

https://user-images.githubusercontent.com/833231/142290416-855bbc4c-dcef-400a-9fdc-aa2466ddf34b.mp4

**Comment by nickytonline** [comment]
@aitchiss, @Ridhwana, and @juliannatetreault, I believe I've addressed all the remaining issues.

File: app/assets/stylesheets/setup-mode.scss
Comment (aitchiss) [reply app/assets/stylesheets/setup-mode.scss:34]: One thing I feel is a bit off with this, is when you click the 'Upload Logo' button, there's kind of an odd flash of this 'Logo is required' text:

https://user-images.githubusercontent.com/20773163/142381576-c1739812-72be-4db2-a625-f01333dfec76.mp4

Also I noticed the padding on the right of the button seems a bit off (maybe the message is causing it to stretch out a bit?)

File: app/javascript/controllers/creator_settings_controller.js
Comment (Ridhwana) [inline app/javascript/controllers/creator_settings_controller.js:1]: @nickytonline will we move this controller to `/admin` in this PR as we discussed yesterday or in a subsequent one?

File: app/assets/stylesheets/setup-mode.scss
Comment (ludwiczakpawel) [reply app/assets/stylesheets/setup-mode.scss:34]: Hey, random idea/thought... And it might be too late for this now... Buuuut gonna say it anyway :D... Do we need to style this input at all? It clearly causes some troubles and a need to do workarounds with label and all kinds of pseudo classes etc... What if we just kept standard style for `input[file]` and show logo preview below it and that's it? I know it won't be that aesthetically pleasing anymore but at least it will be much simpler and it will rely on pretty standard UX patterns.

This is also a thought after all that [conversation around buttons and CTAs](https://forem.team/nickytonline/buttons-links-and-accessbility-587p) too.

File: app/javascript/controllers/creator_settings_controller.js
Comment (nickytonline) [reply app/javascript/controllers/creator_settings_controller.js:1]: Oh yeah. Thanks for the reminder. Will do.

File: app/assets/stylesheets/setup-mode.scss
Comment (nickytonline) [reply app/assets/stylesheets/setup-mode.scss:34]: That's a good point @ludwiczakpawel. The styling was a separate issue but I had styled it so that the preview would fit better and also so the button would look better.

@Ridhwana, are you OK with me unstyling the button for now and we can discuss this further with design as it's more of a application wide consideration?

Comment (Ridhwana) [reply app/assets/stylesheets/setup-mode.scss:34]: Sounds good to me! 👍🏽

Comment (ludwiczakpawel) [reply app/assets/stylesheets/setup-mode.scss:34]: Cool! Apologies from design side: I could probably save you some of the wasted time and several grey hairs if I thought it through more thoroughly earlier. I didn't think it would be that tricky.

File: app/javascript/packs/base.jsx
Comment (nickytonline) [inline app/javascript/packs/base.jsx]: Reverting this for now because for some reason, locally this causes Cypress tests to timeout. This needs to be investigated further in a subsequent PR as it's currently a blocker. cc: @Ridhwana

File: app/javascript/controllers/creator_settings_controller.js
Comment (nickytonline) [reply app/javascript/controllers/creator_settings_controller.js]: Done.

File: app/assets/stylesheets/setup-mode.scss
Comment (nickytonline) [reply app/assets/stylesheets/setup-mode.scss]: No worries. I've put it back to default styling. You can see it in action in the PR description QA section.

**Review by Ridhwana — APPROVED** [review]
🚀 Thank you for all the work on this @nickytonline, it looks great!

**Review by aitchiss — APPROVED** [review]
Nice! 🔥

File: app/views/admin/creator_settings/_form.html.erb
Comment (juliannatetreault) [reply app/views/admin/creator_settings/_form.html.erb:14]: Or even "PNG or JPG" rather than "PNG/JPG" with the forward slash! 😊 

I agree with @ridhwana, though, that this shouldn't hold up the PR and can be adjusted in a separate PR.

Comment (juliannatetreault) [reply app/views/admin/creator_settings/_form.html.erb:14]: Or even "PNG or JPG" rather than "PNG/JPG" with the forward slash! 😊 

I agree with @ridhwana, though, that this shouldn't hold up the PR and can be adjusted in a separate PR.

Comment (juliannatetreault) [reply app/views/admin/creator_settings/_form.html.erb:14]: Or even "PNG or JPG" rather than "PNG/JPG" with the forward slash! 😊 

I agree with @ridhwana, though, that this shouldn't hold up the PR and can be adjusted in a separate PR.

**Review by juliannatetreault — APPROVED** [review]
Thanks so much for tackling this, @nickytonline!! 🚀

File: app/views/admin/creator_settings/_form.html.erb
Comment (nickytonline) [reply app/views/admin/creator_settings/_form.html.erb:14]: I'll mentioned the JPG. Good call. The height of 80px was related to the preview only. I'll omit that part since we'll be auto resizing uploaded logos in a subsequent PR.

**Pull request merged**
